### PR TITLE
Allow custom LocationEngine from NavigationViewOptions

### LINKED ISSUE #957: Feature Request: Allow Custom Location Engines
As far as I can tell, currently the only options for the LocationEngine are using real GPS, or the default simulator.

It is possible to create a new Location Engine, extending com.mapbox.android.core.location.LocationEngine, but there doesn't seem to be a way to set the LocationEngine from outside the library. While in general, these two are enough, there are circumstances where setting our own would be very useful.

I don't think this would be a big change - I think mainly just a change to the LocationEngineConductor, to deactivate the old LocationEngine, set the new one, and activate it.

Ideally, it would also be good if there was an easy way to "transfer" the existing listeners (add them from the old one to the new one), but that can't currently be done directly by the LocationEngineConductor because the "locationListeners" in LocationEngine is protected.

On that note, it would also be good to be able to add a listener to the Location Engine - there are times when having the raw location is important for the UI, rather than the filtered location received in the ProgressChangeListener.

### PR DESCRIPTION
**Pull request by danesfeder**
Closes #957 

Similar to #1232, would love to hear your thoughts / double check for duplicate behavior here @Guardiola31337 @devotaaabel

### CODE REVIEW

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationViewModel.java [hunk 1]
```diff
@@ -254,9 +255,18 @@ private SpeechPlayerProvider initializeSpeechPlayerProvider(boolean voiceLanguag
...
+    LocationEngine locationEngine = options.locationEngine();
+    if (locationEngine == null) {
+      locationEngineConductor.initializeLocationEngine(getApplication(), options.shouldSimulateRoute());
+      locationEngine = locationEngineConductor.obtainLocationEngine();
+    }
+    return locationEngine;
+  }
+
+  private void initializeNavigation(Context context, MapboxNavigationOptions options, LocationEngine locationEngine) {
     navigation = new MapboxNavigation(context, accessToken, options);
```
[the remark this answers is not in the document]
Comment (danesfeder) [reply libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationViewModel.java]: I added a new constructor, as the ones that currently include `LocationEngine` also include `NavigationTelemetry` which I think doesn't make sense to expose

**Comment by danesfeder** [comment]
> I'm wondering why we're removing the listener when calling NavigationTelemetry#updateLocationEngine

@Guardiola31337 this listener is specific to `NavigationTelemetry`

**Comment by danesfeder** [comment]
@Guardiola31337 I made some updates here according to your comments - lmk what you think!

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/location/LocationEngineConductor.java [hunk 2]
```diff
@@ -45,20 +45,35 @@ public LocationEngine obtainLocationEngine() {
     return locationEngine;
   }
 
-  private void initLocationEngine(Context context, boolean simulateRoute) {
+  private void initialize(Context context, LocationEngine locationEngine, boolean simulateRoute) {
     if (simulateRoute) {
-      locationEngine = new ReplayRouteLocationEngine();
+      this.locationEngine = new ReplayRouteLocationEngine();
```
Comment (Guardiola31337) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/location/LocationEngineConductor.java]: Have we considered the option to simulate including a custom `LocationEngine`? Right now, if `options.shouldSimulateRoute()` is `true` we're creating a `ReplayRouteLocationEngine` and ignoring the potential custom `LocationEngine` passed in. Shouldn't be the other way around? I.e. if there's a custom location engine in play we don't rely on the simulation option at all 🤔 Does that make sense? In any case, as both approaches are valid let's make sure we update the docs clarifying this.

**Comment by Guardiola31337** [comment]
@danesfeder 
> this listener is specific to `NavigationTelemetry`

Yeah, I know (sorry if I didn't make myself clear). I was wondering if not "removing + adding" the listener in `NavigationService#updateLocationEngine` (that's what's implemented in `NavigationTelemetry#updateLocationEngine`) would have any implications 🤔 If "re-adding" the listener (without previously removing it) has the same effect let's remove https://github.com/mapbox/mapbox-navigation-android/blob/0a2e7f2f30e405f616342e1dea7c90fa466af358/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationTelemetry.java#L243-L246 completely. Am I missing something here?

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/location/LocationEngineConductor.java [hunk 3]
```diff
@@ -45,20 +45,35 @@ public LocationEngine obtainLocationEngine() {
...
+    if (locationEngine != null) {
+      this.locationEngine = locationEngine;
+    } else {
+      LocationEngineProvider locationEngineProvider = new LocationEngineProvider(context.getApplicationContext());
+      this.locationEngine = locationEngineProvider.obtainBestLocationEngineAvailable();
+      this.locationEngine.setPriority(LocationEnginePriority.HIGH_ACCURACY);
+      this.locationEngine.setFastestInterval(FASTEST_INTERVAL_IN_MILLIS);
+      this.locationEngine.setInterval(INTERVAL_IN_MILLIS);
+    }
+    updateLastLocation();
```
[the remark this answers is not in the document]
Comment (Guardiola31337) [reply libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/location/LocationEngineConductor.java]: Went ahead and opened https://github.com/mapbox/mapbox-navigation-android/pull/1296 to fix what I mentioned ☝️

[the remark this answers is not in the document]
Comment (danesfeder) [reply libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/location/LocationEngineConductor.java]: Awesome thanks @Guardiola31337 ✅

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/location/LocationEngineConductor.java (hunk 2 above)
Comment (danesfeder) [reply libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/location/LocationEngineConductor.java]: @Guardiola31337 yeah let's favor the custom engine before simulation if a custom engine is provided and simulation is enabled

**Comment by danesfeder** [comment]
@Guardiola31337 After some digging on https://github.com/mapbox/mapbox-navigation-android/pull/1257#issuecomment-422340437, I found a 🐛 

The initial setup of `MapboxNavigation#setLocationEngine` would work fine (first time called).  

But if you called this method a second time with a different `LocationEngine`, `NavigationService` would continue to listen to the original engine, because we weren't removing the `NavigationLocationEngineListener` from the first `LocationEngine` (see updated code below).

https://github.com/mapbox/mapbox-navigation-android/pull/1257/files#diff-9f1572844a6e04923accb5f57369797fR22

I updated `MockNavigationActivity` to update the `LocationEngine` when you click the FAB.  The new engine was empty and didn't contain any code, so updates should stop when clicking the FAB:

ezgif com-video-to-gif

This confirms `NavigationTelemetry` is setup correctly and we should remove the listener from the previous engine.

Great catch!

**Pull request merged**